### PR TITLE
feat: create standalone IVF training API in python

### PR DESCRIPTION
**Pull request by westonpace**
This is the start of breaking IVF / PQ index training up into four standalone methods with check pointing in between each method.

### CODE REVIEW

File: python/python/lance/indices.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+                or (
+                    isinstance(field.type, pa.FixedShapeTensorType)
+                    and len(field.type.shape) == 1
+                )
+            ):
+                raise TypeError(
+                    f"Vector column {c} must be FixedSizeListArray "
+                    f"1-dimensional FixedShapeTensorArray, got {field.type}"
+                )
+            if not pa.types.is_floating(field.type.value_type):
```
Comment (eddyxu) [inline python/python/lance/indices.py:247]: We can support `uint8` as well

**Review by eddyxu — APPROVED** [review]

File: python/python/lance/indices.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    A class that represents a trained IVF model.
+    """
+
+    def __init__(self, centroids: pa.Array, distance_type: str):
+        self.centroids = centroids
+        """The centroids of the IVF clusters"""
+        self.distance_type = distance_type
+        """The distance type used to train the IVF model"""
+
+    def num_partitions(self) -> int:
```
Comment (wjones127) [inline python/python/lance/indices.py:28]: Seems like this should be a property:
```suggestion
    @property
    def num_partitions(self) -> int:
```

File: python/python/lance/indices.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        """
+        Train IVF centroids for the given vector column.
+
+        This will run k-means clustering on the given vector column to train the IVF
+        centroids. This is the first step in several vector indices.  The centroids
+        will be used to partition the vectors into different clusters.
+
+        K-means is an iterative algorithm that can be computationally expensive. The
+        accelerator argument can be used to offload the computation to a hardware
+        accelerator such as a GPU or TPU.
```
Comment (wjones127) [inline python/python/lance/indices.py:116]: Is there a progress bar for this? Should we mention the `LANCE_LOG` variable here?

File: python/python/lance/indices.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+from lance.file import LanceFileReader, LanceFileWriter
+
+from .lance import indices
+
+if TYPE_CHECKING:
+    from .dependencies import torch
+
+
+class IvfModel:
```
Comment (wjones127) [inline python/python/lance/indices.py:17]: Not necessary for now, but it would be very cool to be able to get this from an existing index.

**Review by wjones127 — COMMENTED** [review]
Nice work here :)

File: python/python/lance/indices.py (hunk 2 above)
Comment (westonpace) [reply python/python/lance/indices.py:29]: Changed to a property

File: python/python/lance/indices.py (hunk 3 above)
Comment (westonpace) [reply python/python/lance/indices.py:117]: Yes, there is a tqdm progress bar if an accelerator is used.  I started to add one if an accelerator is not used but it ended up being a fair amount of code.  Will open a different PR.

File: python/python/lance/indices.py (hunk 4 above)
Comment (westonpace) [reply python/python/lance/indices.py:17]: Agreed.

**Comment by westonpace** [comment]
I'm going to merge this on green.  I have a few PRs built on top of this one already.

**Pull request merged**